torch/export/_trace.py

```diff
@@ -1510,6 +1604,8 @@ def _is_impure(node):
                     torch.ops.profiler._record_function_enter.default,
                     torch.ops.profiler._record_function_enter_new.default,
                     torch.ops.profiler._record_function_exit._RecordFunction,
+                    torch.ops.aten.detach.default,
+                    builtins.getattr,
```

Comment:
I think this shows up when you are calling extract_val because then you access all inner attributes and check if they are FakeTensor etc.

Reply:
I don't follow though - why do we need to mark these two functions as impure?

My understanding is the main reason we mark some ops as impure is that they has some nonlocal side-effects, and we want to make sure that our graph passes like DCE don't remove them just because they look like dead code.

But both detach and getattr are actual pure functions (they don't have secret side effects)

Reply:
Hmmm so fx DCE calls is_impure function on every node before DCE-ing. These special cased operators will return False (suggesting it is safe to delete). Since we don't want to actually run DCE aggresively, we will by default return True. As a result, only nodes that are marked in this if statement will be deleted.

Reply:
oh I see... it's probably worth a comment for every op that we add to this list (otherwise a year from now someone reading this will have no idea why getattr/detach need to be DCE'd but not other ops). Can you add one describing the situation? (and in particular - that it's not fundamental, and that we can avoid tracing these "dead" getattr/detach calls in principle, but it would require carefully re-writing `track_tensor_tree` to not do any tensor method calls)